http/handlers.go

```diff
@@ -132,64 +155,65 @@ func getMetadata(filter string, model datamodel.DataModel) http.HandlerFunc {
 		if err != nil {
 			l.With("error", err).Info("failed to write response")
 		}
-	}
+	})
 }
 
-func ec2Handler(w http.ResponseWriter, r *http.Request) {
-	if r.Method != "GET" {
-		w.WriteHeader(http.StatusMethodNotAllowed)
-		return
-	}
+func EC2MetadataHandler(logger log.Logger, client hardware.Client) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.Method != "GET" {
+			w.WriteHeader(http.StatusMethodNotAllowed)
+			return
+		}
 
-	logger.Debug("calling ec2Handler ")
-	userIP := getIPFromRequest(r)
-	if userIP == "" {
-		return
-	}
+		logger.Debug("calling ec2Handler ")
+		userIP := getIPFromRequest(r)
+		if userIP == "" {
+			return
+		}
 
-	metrics.MetadataRequests.Inc()
-	l := logger.With("userIP", userIP)
-	l.Info("got ip from request")
-	hw, err := hegelServer.HardwareClient().ByIP(context.Background(), userIP)
-	if err != nil {
-		metrics.Errors.WithLabelValues("metadata", "lookup").Inc()
-		l.With("error", err).Info("failed to get hardware by ip")
-		w.WriteHeader(http.StatusNotFound)
-		return
-	}
+		metrics.MetadataRequests.Inc()
+		l := logger.With("userIP", userIP)
+		l.Info("got ip from request")
+		hw, err := client.ByIP(context.Background(), userIP)
+		if err != nil {
+			metrics.Errors.WithLabelValues("metadata", "lookup").Inc()
+			l.With("error", err).Info("failed to get hardware by ip")
+			w.WriteHeader(http.StatusNotFound)
+			return
+		}
 
-	filter, err := processEC2Query(r.URL.Path)
-	if err != nil {
-		l.With("error", err).Info("failed to process ec2 query")
-		w.WriteHeader(http.StatusNotFound)
-		_, err := w.Write([]byte("404 not found"))
+		filter, err := processEC2Query(r.URL.Path)
 		if err != nil {
-			l.With("error", err).Info("failed to write response")
+			l.With("error", err).Info("failed to process ec2 query")
```

Comment:
Without defining additional error semantics, no, I don't think this is 'path not found'. To disambiguate a 'path not found' from anything else, `err` would need to provide that information.

I can, given these are internal implementation details, define an unexported error and check if we received that as `err` to create 2 different log messages. Alternatively this could be refacftored into 2 stages, `havePath()` and `processPath()` making it much clearer.